### DISCUSSION TITLE
error running z80pack on Raspberry Pi 4

### DISCUSSION THREAD

**Discussion by mmaseda12** [udo-munk/z80pack#392] [General]
I downloaded the latest version and built on the latest version of RaspberryPi 64 bit OS. When I go to run I get the following error:

E (953) IO: can't exec cpmrecv process, compile/install the tools dude
E (8370) IO: can't open pipe auxout

I did install the tools.  Any ideas what I might be doing wrong?

Regards,
Manuel

**Comment by udo-munk** [comment]
This is because the directory where you installed the tools is not included in $PATH.
What I usually do is create a directory bin in my home and then extend the path in the shell's rc file, like this:
export PATH=${HOME}/bin:$PATH

**Comment by mmaseda12** [comment]
Thanks !!ManuelOn Jun 5, 2024, at 10:49 AM, Udo Munk ***@***.***> wrote:﻿
This is because the directory where you installed the tools is not included in $PATH.
What I usually do is create a directory bin in my home and then extend the path in the shell's rc file, like this:
export PATH=${HOME}/bin:$PATH

**Comment by udo-munk** [comment]
You are welcome. How is it working on Pi 4? I still need to get one sometime, but to busy with other things.

**Comment by mmaseda12** [comment]
I just started using but the performance seems to be much better than the 3B I was using.On Jun 5, 2024, at 11:02 AM, Udo Munk ***@***.***> wrote:﻿
You are welcome. How is it working on Pi 4? I still need to get one sometime, but to busy with other things.

**Comment by udo-munk** [comment]
From what I read I would expect that it performs better. How about the FPS for the front panel, does it work with the default 60 or do you have to reduce it?

**Comment by mmaseda12** [comment]
I haven’t built it for Imsai with front panel yet.  Can you point me to the build instructions.  Sent from my iPhoneOn Jun 5, 2024, at 11:34 AM, Udo Munk ***@***.***> wrote:﻿
From what I read I would expect that it performs better. How about the FPS for the front panel, does it work with the default 60 or do you have to reduce it?

**Comment by udo-munk** [comment]
Sure, updated documentation is here: https://github.com/udo-munk/z80pack/tree/master/doc

Start with reading README-frontpanel, then maybe README-wsl, that is an Ubuntu Linux running under WSL on Windows. Don't know what Linux brand one uses on the Pi 4, might need other package names.

**Comment by mmaseda12** [comment]
Thanks,Will build and test in the morning.  Pi uses Debian bookworm.ManuelSent from my iPhoneOn Jun 5, 2024, at 12:33 PM, Udo Munk ***@***.***> wrote:﻿
Sure, updated documentation is here: https://github.com/udo-munk/z80pack/tree/master/doc
Start with reading README-frontpanel, then maybe README-wsl, that is an Ubuntu Linux running under WSL on Windows. Don't know what Linux brand one uses on the Pi 4, might need other package names.

**Comment by udo-munk** [comment]
Good luck. Don't know bookworm, but if you run into problems just ask, will see if we can help.

**Comment by mmaseda12** [comment]
I have everything built and installed and can run Imsai with front panel in 2D mode.  When I try and config for 3D I set a segmentation fault when I click a button on the front panel.  I also tried on a Pi 4 running Ubuntu with the same issue.  Anything I should try?ManuelOn Jun 5, 2024, at 12:39 PM, Udo Munk ***@***.***> wrote:﻿
Good luck. Don't know bookworm, but if you run into problems just ask, will see if we can help.

**Comment by udo-munk** [comment]
Excellent. This likely is caused by bugs in Mesa OpenGL and/or the drivers for the GPU. Not much one can do about this other, than updating all the graphics libs and drivers to the latest version. Whenever I tried to figure a problem on Ubuntu with gdb the segmentation faults happened deep inside some X11 graphics lib. Went away after an update, came again with next update ...

**Comment by mmaseda12** [comment]
I’ll try a few other Linux machines I have around here.On Jun 6, 2024, at 10:17 AM, Udo Munk ***@***.***> wrote:﻿
Excellent. This likely is caused by bugs in Mesa OpenGL and/or the drivers for the GPU. Not much one can do about this other, than updating all the graphics libs and drivers to the latest version. Whenever I tried to figure a problem on Ubuntu with gdb the segmentation faults happened deep inside some X11 graphics lib. Went away after an update, came again with next update ...

**Comment by mmaseda12** [comment]
I tried installing on several versions of Debian and Ubuntu 32 and 64 on both Pi 4 and Pi 5 and still get the segmentation fault in 3D mode.  I even tried on an Intel I7 pc running Mint and get the same error.  In all cases I loaded the most recent drivers and graphics libraries.  I installed on my Intel I7 iMac running Sonoma and it worked perfectly.  Any other suggestions?Regards,ManuelOn Jun 6, 2024, at 10:17 AM, Udo Munk ***@***.***> wrote:﻿
Excellent. This likely is caused by bugs in Mesa OpenGL and/or the drivers for the GPU. Not much one can do about this other, than updating all the graphics libs and drivers to the latest version. Whenever I tried to figure a problem on Ubuntu with gdb the segmentation faults happened deep inside some X11 graphics lib. Went away after an update, came again with next update ...

**Comment by udo-munk** [comment]
I have an older MacBookPro with Intel i5 and a new one with Apple M3, no such problems of course. Currently it works in Windows + WSL with the latest Ubuntu bits, machine has AMD CPU with AMD GPU and current drivers. This is broken from time to time, but right now it works.

Hope you are familiar with gdb, what you can do to locate the problem is this:

Build imsaisim debug version with DEBUG=1 make build
Then rund 'gdb imsaisim' and do what causes the segfault.
When back at the gdb prompt enter command bt (backtrace) and post the output here.

**Comment by mmaseda12** [comment]
ThanksSent from my iPhoneOn Jun 7, 2024, at 11:33 AM, Udo Munk ***@***.***> wrote:﻿
I have an older MacBookPro with Intel i5 and a new one with Apple M3, no such problems of course. Currently it works in Windows + WSL with the latest Ubuntu bits, machine has AMD CPU with AMD GPU and current drivers. This is broken from time to time, but right now it works.
Hope you are familiar with gdb, what you can do to locate the problem is this:
Build imsaisim debug version with DEBUG=1 make build
Then rund 'gdb imsaisim' and do what causes the segfault.
When back at the gdb prompt enter command bt (backtrace) and post the output here.

**Comment by mmaseda12** [comment]
gdb) bt#0  0x00007fffeaac7381 in ?? () from /lib/x86_64-linux-gnu/libLLVM-15.so.1#1  0x00007fffeaac5d82 in llvm::PromoteMemToReg(llvm::ArrayRef<llvm::AllocaInst*>, llvm::DominatorTree&, llvm::AssumptionCache*) ()   from /lib/x86_64-linux-gnu/libLLVM-15.so.1#2  0x00007fffeaeac93d in llvm::SROAPass::runImpl(llvm::Function&, llvm::DominatorTree&, llvm::AssumptionCache&) () from /lib/x86_64-linux-gnu/libLLVM-15.so.1#3  0x00007fffeaeae87f in llvm::SROAPass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) () from /lib/x86_64-linux-gnu/libLLVM-15.so.1#4  0x00007fffecc8b10d in ?? () from /lib/x86_64-linux-gnu/libLLVM-15.so.1#5  0x00007fffea084519 in llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) () from /lib/x86_64-linux-gnu/libLLVM-15.so.1#6  0x00007fffebdca3ad in ?? () from /lib/x86_64-linux-gnu/libLLVM-15.so.1#7  0x00007fffea088401 in llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) ()   from /lib/x86_64-linux-gnu/libLLVM-15.so.1#8  0x00007fffebdca1cd in ?? () from /lib/x86_64-linux-gnu/libLLVM-15.so.1#9  0x00007fffea083129 in llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) ()   from /lib/x86_64-linux-gnu/libLLVM-15.so.1#10 0x00007fffecca746b in LLVMRunPasses ()   from /lib/x86_64-linux-gnu/libLLVM-15.so.1#11 0x00007ffff4fd1a17 in ?? () from /usr/lib/x86_64-linux-gnu/dri/r600_dri.so--Type <RET> for more, q to quit, c to continue without paging--c#12 0x00007ffff4f97cdb in ?? () from /usr/lib/x86_64-linux-gnu/dri/r600_dri.so#13 0x00007ffff4f9b47d in ?? () from /usr/lib/x86_64-linux-gnu/dri/r600_dri.so#14 0x00007ffff4f325bb in ?? () from /usr/lib/x86_64-linux-gnu/dri/r600_dri.so#15 0x00007ffff4f2ba12 in ?? () from /usr/lib/x86_64-linux-gnu/dri/r600_dri.so#16 0x00007ffff4f2be94 in ?? () from /usr/lib/x86_64-linux-gnu/dri/r600_dri.so#17 0x00007ffff4f2c373 in ?? () from /usr/lib/x86_64-linux-gnu/dri/r600_dri.so#18 0x00007ffff4ceb8c5 in ?? () from /usr/lib/x86_64-linux-gnu/dri/r600_dri.so#19 0x00007ffff4cebe15 in ?? () from /usr/lib/x86_64-linux-gnu/dri/r600_dri.so#20 0x00007ffff4ae546c in ?? () from /usr/lib/x86_64-linux-gnu/dri/r600_dri.so#21 0x00007ffff4ae3db0 in ?? () from /usr/lib/x86_64-linux-gnu/dri/r600_dri.so#22 0x00007ffff4c193cd in ?? () from /usr/lib/x86_64-linux-gnu/dri/r600_dri.so#23 0x00005555555d3ea6 in lpSwitch::drawForPick() ()#24 0x00005555555d75a4 in Lpanel::pick(int, int, int, int) ()#25 0x00005555555d1485 in Lpanel::procEvents() ()#26 0x00005555555ce587 in lp_mainloop_thread(void*) ()#27 0x00007ffff7841ac3 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442#28 0x00007ffff78d3850 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81(gdb) On Jun 7, 2024, at 11:33 AM, Udo Munk ***@***.***> wrote:﻿
I have an older MacBookPro with Intel i5 and a new one with Apple M3, no such problems of course. Currently it works in Windows + WSL with the latest Ubuntu bits, machine has AMD CPU with AMD GPU and current drivers. This is broken from time to time, but right now it works.
Hope you are familiar with gdb, what you can do to locate the problem is this:
Build imsaisim debug version with DEBUG=1 make build
Then rund 'gdb imsaisim' and do what causes the segfault.
When back at the gdb prompt enter command bt (backtrace) and post the output here.

**Comment by udo-munk** [comment]
This looks like you started the machine, it displays the frontpanel and then you clicked on some switch:

?? () from /usr/lib/x86_64-linux-gnu/dri/r600_dri.so#23 0x00005555555d3ea6 in lpSwitch::drawForPick() ()[#24](https://github.com/udo-munk/z80pack/issues/24) 0x00005555555d75a4 in Lpanel::pick(int, int, int, int) ()[#25](https://github.com/udo-munk/z80pack/issues/25) 0x00005555555d1485 in Lpanel::procEvents() ()[#26](https://github.com/udo-munk/z80pack/issues/26) 0x00005555555ce587 in lp_mainloop_thread(void*) ()[#27](https://github.com/udo-munk/z80pack/issues/27) 0x00007ffff7841ac3 in start_thread (arg=<optimized out>) at

The frontpanel library wants to redraw the switch because position changed. From there it goes into r600_dri.so, AMD Radeon driver for the GPU.

And then it dies somewhere deep in libLLVM-15.so.1:

gdb) bt#0  0x00007fffeaac7381 in ?? () from /lib/x86_64-linux-gnu/libLLVM-15.so.1#1  0x00007fffeaac5d82 in llvm::PromoteMemToReg(llvm::ArrayRef<llvm::AllocaInst*>, llvm::DominatorTree&, llvm::AssumptionCache*) ()   from /lib/x86_64-linux-gnu/libLLVM-15.so.1#2  0x00007fffeaeac93d in llvm::SROAPass::runImpl(llvm::Function&, 

No idea, sorry. Try again in a few weeks after next update of the Linux stuff.

**Comment by mmaseda12** [comment]
Will do.Best regards,ManuelSent from my iPhoneOn Jun 7, 2024, at 12:30 PM, Udo Munk ***@***.***> wrote:﻿
This looks like you started the machine, it displays the frontpanel and then you clicked on some switch:
?? () from /usr/lib/x86_64-linux-gnu/dri/r600_dri.so#23 0x00005555555d3ea6 in lpSwitch::drawForPick() ()#24 0x00005555555d75a4 in Lpanel::pick(int, int, int, int) ()#25 0x00005555555d1485 in Lpanel::procEvents() ()#26 0x00005555555ce587 in lp_mainloop_thread(void*) ()#27 0x00007ffff7841ac3 in start_thread (arg=) at
The frontpanel library wants to redraw the switch because position changed. From there it goes into r600_dri.so, AMD Radeon driver for the GPU.
And then it dies somewhere deep in libLLVM-15.so.1:
gdb) bt#0  0x00007fffeaac7381 in ?? () from /lib/x86_64-linux-gnu/libLLVM-15.so.1#1  0x00007fffeaac5d82 in llvm::PromoteMemToReg(llvm::ArrayRefllvm::AllocaInst*, llvm::DominatorTree&, llvm::AssumptionCache*) ()   from /lib/x86_64-linux-gnu/libLLVM-15.so.1#2  0x00007fffeaeac93d in llvm::SROAPass::runImpl(llvm::Function&,
No idea, sorry. Try again in a few weeks after next update of the Linux stuff.

**Comment by udo-munk** [comment]
Got my Pi 4B this morning, have setup system and installed OS. Dealer added OS named RB-NOOBS-PI-32GB on SD card. Notes says can also be downloaded from http://www.raspberrypi.org/downloads. Didn't do that but copied it to an unused MicroSD card and installed from and on that. This is some sort of 32 bit Debian Linux, don't know, will see, it is installing latest updates now, the usual procedre...

**Comment by mmaseda12** [comment]
Great !!Let me know if you get the segmentation fault in 3D mode.ManuelSent from my iPhoneOn Jun 11, 2024, at 11:58 AM, Udo Munk ***@***.***> wrote:﻿
Got my Pi 4B this morning, have setup system and installed OS. Dealer added OS named RB-NOOBS-PI-32GB on SD card. Notes says can also be downloaded from http://www.raspberrypi.org/downloads. Didn't do that but copied it to an unused MicroSD card and installed from and on that. This is some sort of 32 bit Debian Linux, don't know, will see, it is installing latest updates now, the usual procedre...

**Comment by udo-munk** [comment]
I get the same result, 3D model segfaults deep in libLLVM, 2D model works really nice, even with 60 fps.

**Comment by mmaseda12** [comment]
Good.  At least it wasn’t anything I was doing wrong.Sent from my iPhoneOn Jun 11, 2024, at 1:28 PM, Udo Munk ***@***.***> wrote:﻿
I get the same result, 3D model segfaults deep in libLLVM, 2D model works really nice, even with 60 fps.

**Comment by sneakywumpus** [comment]
I got my Pi 3B+ out of the drawer and get the same crashes. It runs when using "LIBGL_ALWAYS_SOFTWARE=1 ./cpm22".

**Comment by mmaseda12** [comment]
What’s also interesting is I get the same error on an Intel pc running the latest version of Linux Mint. ManuelOn Jun 12, 2024, at 10:32 AM, sneakywumpus ***@***.***> wrote:﻿
I got my Pi 3B+ out of the drawer and get the same crashes. It runs when using "LIBGL_ALWAYS_SOFTWARE=1 ./cpm22".

**Comment by sneakywumpus** [comment]
It runs fine on my PC with Arch Linux and Mesa 24.1.1

**Comment by mmaseda12** [comment]
I’ll load a PC with Arch and see what I get.Sent from my iPhoneOn Jun 12, 2024, at 10:39 AM, sneakywumpus ***@***.***> wrote:﻿
It runs fine on my PC with Arch Linux and Mesa 24.1.1

**Comment by udo-munk** [comment]
LIBGL_ALWAYS_SOFTWARE=1
This lets Mesa using the software renderer regadless if there is some GPU driver or not. And with software rendering it is working, slow off course and LED's are not drawn correct, indicates a bug in the GPU driver code.

**Comment by udo-munk** [comment]
Hm, not using a GPU driver:

udo@raspberrypi:~/work/z80pack $ glxinfo | grep render
direct rendering: Yes
    GLX_MESA_query_renderer, GLX_MESA_swap_control, GLX_OML_swap_method,
    GLX_MESA_query_renderer, GLX_OML_swap_method, GLX_SGIS_multisample,
Extended renderer info (GLX_MESA_query_renderer):
OpenGL renderer string: llvmpipe (LLVM 11.0.1, 128 bits)
    GL_ARB_conditional_render_inverted, GL_ARB_conservative_depth,
    GL_NV_conditional_render, GL_NV_copy_image, GL_NV_depth_clamp,
    GL_ARB_conditional_render_inverted, GL_ARB_conservative_depth,
    GL_NV_conditional_render, GL_NV_copy_depth_to_color, GL_NV_copy_image,
    GL_EXT_render_snorm, GL_EXT_robustness, GL_EXT_sRGB_write_control,
    GL_MESA_shader_integer_functions, GL_NV_conditional_render,
    GL_OES_element_index_uint, GL_OES_fbo_render_mipmap,

**Comment by sneakywumpus** [comment]
I get this on my 3B+:
```
thomas@raspi:~ $ glxinfo -B
name of display: :0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Broadcom (0x14e4)
    Device: VC4 V3D 2.1 (0xffffffff)
    Version: 23.2.1
    Accelerated: yes
    Video memory: 907MB
    Unified memory: yes
    Preferred profile: compat (0x2)
    Max core profile version: 0.0
    Max compat profile version: 2.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 2.0
OpenGL vendor string: Broadcom
OpenGL renderer string: VC4 V3D 2.1
OpenGL version string: 2.1 Mesa 23.2.1-1~bpo12+rpt3
OpenGL shading language version string: 1.20

OpenGL ES profile version string: OpenGL ES 2.0 Mesa 23.2.1-1~bpo12+rpt3
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 1.0.16
```

**Comment by mmaseda12** [comment]
My Mint PC is also running Mesa 23.  Can it be updated to Mesa 24?Sent from my iPhoneOn Jun 12, 2024, at 11:37 AM, sneakywumpus ***@***.***> wrote:﻿
I get this on my 3B+:
***@***.***:~ $ glxinfo -B
name of display: :0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
Vendor: Broadcom (0x14e4)
Device: VC4 V3D 2.1 (0xffffffff)
Version: 23.2.1
Accelerated: yes
Video memory: 907MB
Unified memory: yes
Preferred profile: compat (0x2)
Max core profile version: 0.0
Max compat profile version: 2.1
Max GLES1 profile version: 1.1
Max GLES[23] profile version: 2.0
OpenGL vendor string: Broadcom
OpenGL renderer string: VC4 V3D 2.1
OpenGL version string: 2.1 Mesa 23.2.1-1~bpo12+rpt3
OpenGL shading language version string: 1.20
OpenGL ES profile version string: OpenGL ES 2.0 Mesa 23.2.1-1~bpo12+rpt3
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 1.0.16

**Comment by udo-munk** [comment]
Ah, forgot, I disabled the driver in /boot/conf.txt, that is why it is running always with software renderer.
Enabled again and now I get the Broadcom VC4 V3D driver.

**Comment by udo-munk** [comment]
One could try the current dev driver:

The driver code for both GPU models is available in the [Mesa upstream repository](https://gitlab.freedesktop.org/mesa/mesa/). The codename for the VideoCore IV driver is VC4, and the codename for the VideoCore VI driver is V3D. There are no downstream repositories – all development happens directly upstream, which has a number of benefits for end users:

It is relatively easy for the more adventurous users to experiment with development builds of the driver.

**Comment by udo-munk** [comment]
I tried one more thing. I noticed that the OS I installed was the 32-bit version, but I got a Pi with 8 GB memory. For this one needs a 64-bit OS to take full advantage, and that should avoid several of the usual problems to access memory > 4 GB with the 32 bits. Machine is more performant with the 64-bit OS, but the problem with the GPU driver still is there, so we have to wait for a better GPU driver.

**Comment by mmaseda12** [comment]
﻿I wasn’t able to get Arch installed but I did install Tumbleweed with Mesa 24.0.8 and I still get the segmentation fault in 3D mode.  Perhaps 24.1.1 fixes that.ManuelOn Jun 12, 2024, at 10:39 AM, sneakywumpus ***@***.***> wrote:﻿
It runs fine on my PC with Arch Linux and Mesa 24.1.1

**Comment by RanchoHam** [comment]
I'm running RPI4's (Bullseye) with good performance with only 2D front panels.  (BTW I don't really see the point of 3D, so 2D works for me.)  I have been looking through 1.37 for months to find why I wasn't getting the web server to work.  When I upgraded to 1.38, I noticed a message that the `Web server builtin, but disabled` which explains things.  But how does one turn it on?

**Comment by sneakywumpus** [comment]
Use option `-n` to enable to web server. Port number is 8080, which can be changed in conf/system.conf (ns_port).

**Comment by udo-munk** [comment]
Well, the point of 3D is, that John who wrote the frontpanel library, is a sophisticated OpenGL programmer, and wanted to show off some 3D modelling. Even the frontpanel is reflective plexi glas and seems to show the environment. It is nice to look at, but for using the machines the 2D view is more comfortable of course, and all this is configurable for the end user.

In releases 1.36 & 1.37 the web server was not compiled in by default, it was a compile time option in sim.h of the machines, and one had to enable it there. With 1.38 the web server always is compiled into the machines, but by default it is disabled. Enable with running the machines with -n.

**Reply by RanchoHam** [reply]
Thanks @sneakywumpus and @udo-munk for the -n hint.  I tried it in both the imsai and cromemco directories and it looks good.  

But the LIB: only contains the disks in the ./disks directory not the ./disks/library directory.  Further examination of the SYS: shows the **DISKSDIR=(null)**. I have Dave's Cromemco Z-1; and it shows **DISKSDIR=/sdcard/cromemco/disks**, so  I tried `export DISKDIR=<full path to ./disks/library>`, but that did not help.  What am I doing wrong?

**Comment by udo-munk** [comment]
The used disk directories are compiled in and cannot be changed with environment variables. For the PC version you can define another path with disk images with the option -r <path>. However, the web lib manager will only look in ./disks, so copy the disk images you wish to use into this directory.

**Reply by RanchoHam** [reply]
Thanks @udo-munk, I have imsai and cromemco sims working on a Windows WSL2/Ubuntu 22 installation (although it seg faults on exit) and on a Raspberry Pi 4/Raspberry Pi OS (Bullseye 64) with normal closure.

**Comment by udo-munk** [comment]
Here is how you can run the z80pack 3D models on the Raspberry Pi, by not using their broken graphics drivers, but using a working remote one. For example I am using Linux Ubuntu in WSL2 running on a machine with AMD hardware (RYZEN 7 with AMD RADEON). On the Windows system I am using MobaXTerm and in the setting make sure that X11 -> OpenGL acceleration is set to Hardware.
With this now ssh to your Raspberry Pi, make sure that X11 tunneling is configured for the session. On Pi command line enter:

export LIBGL_ALWAYS_INDIRECT=1

Now the frontpanel gets drawn by your PC GPU hardware with it's drivers.

**Comment by udo-munk** [comment]
Tested z80pack with SDL2 graphics as added by Thomas. Unlike the X11 version the 3D IMSAI model works OK. Performance is good and box looks same nice as the X11 version on larger workstations.

**Reply by mmaseda12** [reply]
Great, I’m out of town but I’ll try on Sunday when I get home.

**Comment by mmaseda12** [comment]
Did a clean install on a Pi4 with Ubuntu and 3D works perfectly.

**Comment by udo-munk** [comment]
Excellent. And also with sound, I have that enabled via the HDMI output to the monitor, which also has speakers builtin, this also works on the Pi systems.

**Reply by mmaseda12** [reply]
How do I enable sound?  I already have Ubuntu configured to send sound over HDMI.

**Comment by udo-munk** [comment]
I use Raspberry Pi OS 64bit, this is Debian based and the GUI probably is different in the Ubuntu distribution. However, there should be a system setting, that allows to route audio to HDMI or the audio jack. I set it to HDMI, set audio volume to 50% and checked with YouTube if sound is working.

In the latest z80pack sources audio is enabled for the switches, but disabled for the cooling fan. The configuration for the sounds is in <machine>sim/conf/panel.conf.

**Reply by mmaseda12** [reply]
Must be a Ubuntu issue.  Works fine on a Pi running PI OS 64.  But doesn’t work on a Pi running Ubuntu 64 24.10.  I have confirmed that that the sound is configured properly to work over HDMI.  I can use YouTube and play WAV files.

**Reply by mmaseda12** [reply]
Problem solved.  Ubuntu 24.10 uses Rhythmbox as the default media player.  Pi OS uses VLC.  I removed Rhythmbox on Ubuntu and replaced it with VLC and everything works now.  Perhaps there is a problem between SDL2 and Rhythmbox.

**Comment by udo-munk** [comment]
Today I have some news for those using z80pack on Raspberry Pi computers :-)
Today I updated my system and the huge updates included low level graphics drivers and mesa libraries. So I just tried the IMSAI 3D X11 model with a fresh build and guess what, it now works. Looks like they finaly fixed these issues with the 3D graphics.

**Reply by mmaseda12** [reply]
Great, I’ll do a fresh install on a Pi 5 this weekend and check it out.

**Comment by mmaseda12** [comment]
Worked fine.  Thanks !!

**Comment by udo-munk** [comment]
Excellent, glad it works on your machine too.
